# введение в Pandas
## ghghghgh
### Базовый набор методов и приемов в Pandas

описалово

ноут отредактирован 16.11.2022  
[ссылка на курс](https://stepik.org/lesson/681964/) 

### 1. Импорт библиотек

In [20]:
import pandas as pd
import numpy as np

### 2. Загрузка данных

In [21]:
path = 'rides_info_short.csv'
rides_info = pd.read_csv(path)

***[pd.read_csv()](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html)*** - cчитывает данные в DataFrame из файла csv (файл с разделителями). У метода очень много параметров с которыми нужно ознакомится самостоятельно.  

### 3. Предварительный осмотр данных

In [22]:
rides_info.sample(5)

car_id ride_id  speed_avg    rating  ride_duration  ride_cost  \
5698   A17304887K     D1E  35.000000  3.049037       4.628990   4.852303   
15985  A66878295e     N1Z  67.271229  3.242520       4.835625   3.455306   
893    A11000629j     u1Q  34.000000  4.369930       5.319750   4.013269   
7066   A18490474Q     b1q  33.000000  4.492565       4.763158   3.919789   
7974   A18948234K     A1Q  41.000000  5.499077       4.258583   6.195338   

       stop_times  
5698            2  
15985           0  
893             0  
7066            2  
7974            2

***[df.sample()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sample.html)*** - возвращает случайную выборку строк датафрейма. Параметр n - кол-во выводимых строк.  

здесь и далее под df подразумевается любой датафрейм к которому вы применяете метод  

In [23]:
# часто используют метод head(), но лучше первый раз смотреть на датафрейм через sample()
# потомучто так вы имеете возможность смотреть на разные выборки и это может быть полезно, чтобы увидеть какие-нибудь баги

rides_info.head()

car_id ride_id  speed_avg    rating  ride_duration  ride_cost  \
0  A10077085b     p1V  36.000000  4.806122       4.650555   3.721202   
1  A10077085b     F1h  41.000000  5.827247       4.663927   7.126641   
2  A10077085b     m1j  40.581614  3.497089       4.689062   6.323878   
3  A10077085b     a1n  29.000000  4.896888       5.212562   5.326560   
4  A10077085b     B1N  31.000000  4.151057       3.430144   4.145335   

   stop_times  
0           1  
1           1  
2           1  
3           0  
4           2

***[df.head()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.head.html)*** - возвращает первые строки датафрейма. По умолчаю выводит 5 строк, но можно изменить параметр n и вывести больше.  

***Выводы:***
- в целом данные загрузились 
- названия столбцов из первой строки загружены
- индексы первого столбца выглядят адекватно
- наличие индексов, которые нужно удалить не обнаружено
- по случайным пяти строкам наличие каких-либо ошибок не обнаружено

### 4. Предварительный анализ данных

In [24]:
# в нашем случае датафрейм небольшой, но он может быть большим. Мы разберем методы библиотеки Pandas
# на этом небольшом примере, которые вы по аналогии самостоятельно сможете применить к другим датасетам

rides_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18000 entries, 0 to 17999
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   car_id         18000 non-null  object 
 1   ride_id        18000 non-null  object 
 2   speed_avg      18000 non-null  float64
 3   rating         18000 non-null  float64
 4   ride_duration  18000 non-null  float64
 5   ride_cost      18000 non-null  float64
 6   stop_times     18000 non-null  int64  
dtypes: float64(4), int64(1), object(2)
memory usage: 984.5+ KB


***[df.info()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.info.html)*** - Выводит краткую сводку информацию по датафрейму.  

***Выводы:***
- в датафрейме 7 столбцов
- два из них имеют тип object
- четыре из них имеют тип float64
- один столбец имеет тип int64
- всего в датафрейме 18000 строк
- пропусков нет (во всех столбцах кол-во ненулевых значений равно 18000)
- объем занимаемый датафреймом 984.5+ KB (эта информация может быть полезна, чтобы заранее представить какие операции с датафреймом вы сможете выполнять с учетом доступной вам оперативной памяти)

In [25]:
# иногда кол-во столбцов очень большое, так что при выводе они все не помещаются в стандартный вывод через df.info(),
# тогда можно воспользоваться атрибутом для просмотра индексов названий столбцов
rides_info.columns

Index(['car_id', 'ride_id', 'speed_avg', 'rating', 'ride_duration',
       'ride_cost', 'stop_times'],
      dtype='object')

  

***[df.columns](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.columns.html)*** - Выводит метки столбцов датафрейма.  

In [26]:
# гораздо нагляднее и удобнее работать со списком названий столбцов
column_names = list(rides_info.columns)
column_names, len(column_names)

(['car_id',
  'ride_id',
  'speed_avg',
  'rating',
  'ride_duration',
  'ride_cost',
  'stop_times'],
 7)

***Выводы:***
- 7 столбцов
- все названия столбцов в нижнем регистре и не содержат спецсимволов (значит их не нужно переименовывать, чтобы обращаться к ним через точку)


### 5. Предварительный анализ числовых столбцов

In [28]:
rides_info.describe()

speed_avg        rating  ride_duration     ride_cost    stop_times
count  18000.000000  18000.000000   18000.000000  18000.000000  18000.000000
mean      38.806994      4.399343       4.923314      4.931399      0.808167
std       11.462446      0.998470       0.993852      0.996623      0.832976
min       15.000000      0.751504       0.801373      1.315689      0.000000
25%       32.000000      3.727872       4.244142      4.266755      0.000000
50%       37.000000      4.391036       4.928637      4.932504      1.000000
75%       41.000000      5.073972       5.593254      5.595347      1.000000
max      103.911531      7.954961       8.895110      9.444028      5.000000

***[df.describe()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html)*** - Выводит описательные статистики по числовым столбцам датафрейма. Описательные статистики включают в себя
- count - кол-во не пустых значений
- mean - среднее арифметическое
- std - дисперсия
- min, max - минимум и максимум
- 25%, 50%, 75% - соотвествующие квартили  
- 
!!!внимание!!! из статистик автоматически исключаюися NaN значения, если они есть.  

***Некоторые выводы:***
- описание столбцов датафрема (speed_avg - средняя скорость по маршруту такси, rating - рейтинг полученный водителем за поездку, ride_duration - продолжительность поездки, ride_cost - стоимость поездки, stop_times - кол-во остановок)
- средняя скорость движения по маршруту по всем поездкам около 40, а максимальная средняя скорость 103. Вероятно этот столбец емеет размерность км/час
- как минимум в половине случаев во время поездки была одна остановка
- максимальный рейтинг около 8 минимум чуть ниже единицы. Более половины всех рейтингов с оценками от 3 до 5. Вероятно размерность этого столбца баллы.
- сделать однозначные выводы о размерностях продолжительности и стоимости поездок нельзя, вероятно они были модифицированы и для этого требуется дополнительный анализ и визуализация данных 

### 6. Работа со столбцами и строками

In [34]:
# выше мы рассмотрели методы которые можно применить ко всему датафрейму.  
# далее нам нужно будет перейти к анализу столбцов типа object.  
# И прежде чем, это сделать нам необходимо разобраться как работать со столбцами и строками  

# ниже приводяться несколько способов обращения к столбцам
rides_info['car_id']

0        A10077085b
1        A10077085b
2        A10077085b
3        A10077085b
4        A10077085b
            ...    
17995    A89842534W
17996    A89842534W
17997    A89842534W
17998    A89842534W
17999    A89842534W
Name: car_id, Length: 18000, dtype: object

In [35]:
rides_info.car_id

0        A10077085b
1        A10077085b
2        A10077085b
3        A10077085b
4        A10077085b
            ...    
17995    A89842534W
17996    A89842534W
17997    A89842534W
17998    A89842534W
17999    A89842534W
Name: car_id, Length: 18000, dtype: object

In [33]:
rides_info.loc[:, 'car_id']

0        A10077085b
1        A10077085b
2        A10077085b
3        A10077085b
4        A10077085b
            ...    
17995    A89842534W
17996    A89842534W
17997    A89842534W
17998    A89842534W
17999    A89842534W
Name: car_id, Length: 18000, dtype: object

In [40]:
rides_info.iloc[:,0]

0        A10077085b
1        A10077085b
2        A10077085b
3        A10077085b
4        A10077085b
            ...    
17995    A89842534W
17996    A89842534W
17997    A89842534W
17998    A89842534W
17999    A89842534W
Name: car_id, Length: 18000, dtype: object

На этом этапе важно разобратся с понятием Series

<img src = '../images/pandas_2.jpg'>

In [65]:
# теперь, если мы хотим добавить новый столбец cheap_price - характеристика дешевой цены поездки:
# если цена в первом 25% квартиле (то есть меньше 4.26) то это будет дешевая цена 'Yes', иначе 'No'


# как нам вывести часть датафрейма с дешевыми ценами, попробуем просто записать это условие и вывести его
rule_cheap = rides_info.ride_cost < 4.26
rule_cheap[:5]

0     True
1    False
2    False
3    False
4     True
Name: ride_cost, dtype: bool

Если вы посмотрите выше на первые пять строк датафрейма, то вы увидите, что строки с индексами 0 и 4 (первая и пятая строка) имеют цену ниже 4.26. Таким образом rule_cheap содержит True в значениях, удовлетворяющих условию.

In [52]:
# мы можем создать новый датафрейм только с дешевыми ценами
rides_cheap_price = rides_info[rule_cheap]
rides_cheap_price.head()

car_id ride_id  speed_avg    rating  ride_duration  ride_cost  \
0   A10077085b     p1V  36.000000  4.806122       4.650555   3.721202   
4   A10077085b     B1N  31.000000  4.151057       3.430144   4.145335   
10  A10077085b     H1s  52.416858  3.465121       5.061486   3.873964   
12  A10077085b     B1d  38.000000  5.865320       4.405884   3.967056   
16  A10077085b     a1d  32.091747  3.723012       5.052910   4.175643   

    stop_times  
0            1  
4            2  
10           0  
12           1  
16           1

In [51]:
# 
len(rides_cheap_price)

4465

ссылка на лен

In [66]:
# мы можем создать новый столбец с дешевыми ценами
rides_info.loc[rule_cheap, 'cheap_price'] = 'Yes'
rides_info.head()

car_id ride_id  speed_avg    rating  ride_duration  ride_cost  \
0  A10077085b     p1V  36.000000  4.806122       4.650555   3.721202   
1  A10077085b     F1h  41.000000  5.827247       4.663927   7.126641   
2  A10077085b     m1j  40.581614  3.497089       4.689062   6.323878   
3  A10077085b     a1n  29.000000  4.896888       5.212562   5.326560   
4  A10077085b     B1N  31.000000  4.151057       3.430144   4.145335   

  cheap_price  stop_times  
0         Yes           1  
1         NaN           1  
2         NaN           1  
3         NaN           0  
4         Yes           2

In [67]:
# так неудобно сравнивать два столбца давайте передвинем cheap_price к ride_cost
rides_info = rides_info[column_names[:-1]+['cheap_price']+['stop_times']]
rides_info.head()

car_id ride_id  speed_avg    rating  ride_duration  ride_cost  \
0  A10077085b     p1V  36.000000  4.806122       4.650555   3.721202   
1  A10077085b     F1h  41.000000  5.827247       4.663927   7.126641   
2  A10077085b     m1j  40.581614  3.497089       4.689062   6.323878   
3  A10077085b     a1n  29.000000  4.896888       5.212562   5.326560   
4  A10077085b     B1N  31.000000  4.151057       3.430144   4.145335   

  cheap_price  stop_times  
0         Yes           1  
1         NaN           1  
2         NaN           1  
3         NaN           0  
4         Yes           2

In [68]:
# нам надо заполнить пустые значения NaN значениями 'No'
# мы воспользуемся методом заполнения пропусков
rides_info.cheap_price.fillna('No', inplace=True)
rides_info.head()

car_id ride_id  speed_avg    rating  ride_duration  ride_cost  \
0  A10077085b     p1V  36.000000  4.806122       4.650555   3.721202   
1  A10077085b     F1h  41.000000  5.827247       4.663927   7.126641   
2  A10077085b     m1j  40.581614  3.497089       4.689062   6.323878   
3  A10077085b     a1n  29.000000  4.896888       5.212562   5.326560   
4  A10077085b     B1N  31.000000  4.151057       3.430144   4.145335   

  cheap_price  stop_times  
0         Yes           1  
1          No           1  
2          No           1  
3          No           0  
4         Yes           2

In [70]:
# теперь давайте приведем датафрейм к исходному виду заодно научимся удалять ненужные нам столбцы


**[df.fillna()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.fillna.html)** - Заполняет значения NA/NaN.  
Параметр inplace=True заполняет в датафрейме без необходимости копирования. По умолчанию inplace=False

### 7. Предварительный анализ столбцов типа object

In [41]:
rides_info.car_id.nunique()

100

In [58]:
rides_info.car_id.unique()

array(['A10077085b', 'A10348902N', 'A10457102q', 'A10976678f',
       'A11000629j', 'A11289028f', 'A11516291P', 'A12154061t',
       'A12497993q', 'A12509261Q', 'A13037607O', 'A13388914L',
       'A13399615r', 'A13860178h', 'A14158064f', 'A15097660Y',
       'A15293803l', 'A15819179M', 'A15892333D', 'A16089058W',
       'A16182928S', 'A16214802X', 'A16226657I', 'A16299577x',
       'A16371633V', 'A16528259W', 'A16827262J', 'A16899458Q',
       'A17083016y', 'A17146817r', 'A17284553a', 'A17304887K',
       'A17388761D', 'A17486515d', 'A17917797A', 'A18007714A',
       'A18128039l', 'A18347531o', 'A18431168S', 'A18490474Q',
       'A18545672F', 'A18795645R', 'A18824375A', 'A18826686K',
       'A18948234K', 'A19038463m', 'A19118391L', 'A19836425I',
       'A20293658W', 'A20428618x', 'A20849766t', 'A21152709y',
       'A21177363z', 'A21183563Q', 'A21532200L', 'A22058082P',
       'A22175034X', 'A22245696f', 'A22319160W', 'A22464969m',
       'A22505352r', 'A22845474A', 'A23037659n', 'A2643

In [59]:
rides_info.car_id.value_counts()

A10077085b    180
A26437578J    180
A38864712f    180
A38653666j    180
A37755369F    180
             ... 
A17284553a    180
A17146817r    180
A17083016y    180
A16899458Q    180
A89842534W    180
Name: car_id, Length: 100, dtype: int64

In [60]:
rides_info.car_id.value_counts().describe()

count    100.0
mean     180.0
std        0.0
min      180.0
25%      180.0
50%      180.0
75%      180.0
max      180.0
Name: car_id, dtype: float64

In [61]:
rides_info.ride_id.nunique()

2701

In [62]:
rides_info.ride_id.unique()

array(['p1V', 'F1h', 'm1j', ..., 'j1U', 'f1G', 'n1Y'], dtype=object)

In [63]:
rides_info.ride_id.value_counts()

L1C    17
K1P    16
k1L    16
r1t    15
P1k    15
       ..
M1O     1
r1p     1
E1e     1
G1t     1
R1R     1
Name: ride_id, Length: 2701, dtype: int64

### 8. Простые методы генерации новых признаков

В этом разделе мы попробуем сгенерировать новые признаки используя простые приемы Pandas и Numpy

Мы рассмотрели основные методы и примемы для работы с Pandas.  
Далее вас ждут упражнения по этим методам, для закрепления материала. Удачи!


Следующий шаг в анализе данных датафрейма предполагает применение более продвинутых методов.  
